### ISSUE TITLE
Global hotkey not supported on Wayland

### ISSUE THREAD

**Issue by FredHaa** [warpdotdev/Warp#4800]
### Describe the bug

After newest update Global hotkey no longer works with Wayland.
It worked somewhat before (The hotkey only worked when some windows were in focus like Slack and VSCode, not Firefox and Nautilus).

### To reproduce

Go to setting -> features -> keys
Global hotkey feature greyed out.

### Screenshots

![Screenshot from 2024-04-21 18-33-40](https://github.com/warpdotdev/Warp/assets/6875946/88fb4b79-8424-4fed-a967-ab3762b0ac58)

### Operating system

Linux

### Operating system and version

Archlinux

### Shell Version

zsh 5.9

### Current Warp version

v0.2024.04.16.08.02.stable_00

### Regression

Yes, this bug started recently or with an X Warp version

### Does this block you from using Warp daily?

No

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by acarl005** [comment]
Hi @FredHaa. Unfortunately, the Wayland protocol simply doesn't have the functionality to support this feature. Although [xdg-desktop-portal](https://github.com/flatpak/xdg-desktop-portal/pull/711) _does_ have global hotkeys, Wayland will not allow, for example, an application to bring its window to the front. Currently, the closest thing Wayland allows is for a window to ["ask" for activation](https://wayland.app/protocols/xdg-activation-v1), which won't actually show the window in most desktop environments. As such, if Warp acts as a Wayland client, it cannot support this feature anymore.

When Warp acts as an X11 client (through Xwayland), this feature becomes _kind of_ possible due to the backwards compatilibity. As you've noticed, it is finnicky, since we can only capture global hotkeys when another X11 client is focused, such as Chrome or other Chromium-based apps. If you want to get back the old behavior, you can toggle this "Use X11 for window management" option in the settings.

![Screenshot from 2024-04-22 15-45-03](https://github.com/warpdotdev/Warp/assets/8334252/e3af9607-9f54-4a34-af54-dc9ee90de728)

There are 2 alternatives as well:

1. Configure a hotkey with your window manager. This is highly dependent on your desktop environment. But, there should be a way to bind a keystroke to show Warp.
2. Switch your desktop session to use X11 instead of Wayland. In this case, Warp's own global hotkey should work perfectly. However, it comes with the drawback that you have to go back to X11 which we don't necessarily recommend. It would come with other limitations w.r.t. scaling, frame rates, and any other baggage with X11.

**Comment by FredHaa** [comment]
Sounds good with the option to run with XWayland.

Even better would be to fix the problem similar to how [Guake does it](https://wiki.archlinux.org/title/Guake#Toggling_Guake_visibility_does_not_work_(Wayland)). I.e. implementing a command `warp-toggle`  which would trivialize setting up the window manager hotkey as you suggest as the first alternative.

**Comment by FredHaa** [comment]
For reference, Guake uses DBus to send a message to toggle the window:

```python
def toggle_guake_by_dbus():
    import dbus  # pylint: disable=import-outside-toplevel

    try:
        bus = dbus.SessionBus()
        remote_object = bus.get_object("org.guake3.RemoteControl", "/org/guake3/RemoteControl")
        print("Sending 'toggle' message to Guake3")
        remote_object.show_hide()
    except dbus.DBusException:
        pass
```

**Comment by acarl005** [comment]
Ah thanks for the pointer. I'll dig into the Guake method and see what Warp can do herel.

**Comment by HrBingR** [comment]
Just adding to this, I would love the guake method; either being able to make use of the dbus to write my own command to toggle warp, or just a 'warp-toggle' command being exposed that I can setup a Gnome shortcut for would be amazing. Currently I have to use 'warp-terminal' which works okay, but it only launches warp (rather than toggles) and will just keep launching new copies. 

Sadly Gnome also doesn't have a setting for legacy x11 window management so that's what I'm currently stuck with, sadly.

**Comment by FredHaa** [comment]
Alright, thanks for the update

**Comment by masch** [comment]
Hi! Is there any update on this issue, or is there a workaround to make it work?
Thanks!

**Comment by acarl005** [comment]
@masch Workarounds are either to use an X11 window manager or use a Wayland window manager which lets you set up global hotkeys to show a window.

**Comment by roxi3906** [comment]
+1, switched to xorg to use this feature. But we'd like to see some progress on wayland.

**Comment by iamchriswick** [comment]
# 🚨 Fix for Global Hotkey Issues on Linux with Wayland 🚨

Hey Warp team! 👋 

I've been experiencing the same Global Hotkey issues that many Linux users face with Wayland, and I found how **Ulauncher solved this exact problem**. Here's their successful approach that Warp could implement:

## The Problem 
Warp's Global Hotkey feature doesn't work reliably on Wayland because:
- Wayland has stricter security models for global shortcuts
- The built-in hotkey registration fails when triggered from certain windows (terminals, system settings, etc.)

## Ulauncher's Solution ✅
Instead of relying on internal global hotkey registration, Ulauncher provides:

1. **A dedicated toggle command**: `ulauncher-toggle` 
2. **User instructions** to:
   - Use `wmctrl` (for window focus activation) - *already installed by default on Fedora and most modern distros*
   - Set Ulauncher's internal hotkey to something unused
   - Create a custom system shortcut in GNOME/KDE that runs `ulauncher-toggle`

This bypasses Wayland's limitations by using the system's native shortcut management.

## Proposed Solution for Warp 🎯

**Add a `warp-terminal-toggle` command** that:
- Shows the dedicated hotkey window if hidden
- Hides the dedicated hotkey window if visible  
- Uses `wmctrl` for proper window focus (already available on most systems)

Then update the documentation to guide users:
1. **No installation needed** - `wmctrl` is already installed on Fedora and most modern distros
2. Set Warp's internal Global Hotkey to "disabled" or unused key
3. Create system shortcut: `Settings > Keyboard > Custom Shortcuts`
4. Command: `warp-terminal-toggle`
5. Set their preferred hotkey

## Why This Works
- ✅ Bypasses Wayland's global hotkey limitations
- ✅ Uses system-native shortcut handling 
- ✅ Provides consistent behavior across different Wayland compositors
- ✅ Maintains the same UX as the current Global Hotkey feature
- ✅ **No additional dependencies** - leverages existing system tools
- ✅ **Follows the proven `ulauncher-toggle` pattern**

## Current Status
Right now, there's **no equivalent to `ulauncher-toggle`** in Warp - the CLI only supports basic options without any window management commands.

This seems like a relatively straightforward addition that would solve the Wayland hotkey issues for good! 🚀

## References
- Ulauncher Wayland solution: https://github.com/Ulauncher/Ulauncher/wiki/Hotkey-In-Wayland
- Warp Global Hotkey docs: https://docs.warp.dev/terminal/windows/global-hotkey

What do you think? Would this approach work for Warp's architecture? Happy to help test or provide more details! 

**Environment:** Fedora 42, Wayland, GNOME

**Comment by acarl005** [comment]
Hi @iamchriswick thanks for engaging. I appreciate your suggestions. My feedback is that I'm hesitant to depend on `wmctrl` for window focus. Although it may be included in current versions of some distros, we cannot assume this will always be the case. `wmctrl` is based on X11 and given the effort to remove X11 support from DE's like GNOME, it might get dropped as an included component. As for global hotkey registration, I'm also hesitant to use any DE-specific methods. There are a lot of users outside of GNOME/KDE and it's impractical to try to target every popular DE/WM. In general we prefer to stick to standards that apply to all Linux desktop, e.g. Wayland/X11, XDG, portals, d-bus etc. If Tauri ends up implementing something here we'll likely pull in their approach.  tauri-apps/global-hotkey#28

**Comment by ElijahLynn** [comment]
I got this working on Wayland + Gnome by vibe coding a custom Gnome Extension - https://github.com/ElijahLynn/warp-visor, it hardcodes the shortcuts (top and bottom visor) and has some bugs but I use it as my daily driver and it has made my Warp experience enjoyable again. If Warp finds it useful, you are welcome to adopt it (I can transfer to you), or rewrite it etc. 

It could do with being published as a real Gnome Extension and make it easy to install.

**Comment by acarl005** [comment]
@ElijahLynn Thanks for the offer. We would prefer not to maintain a bunch of DE-specific methods.

**Comment by ElijahLynn** [comment]
> [@ElijahLynn](https://github.com/ElijahLynn) Thanks for the offer. We would prefer not to maintain a bunch of DE-specific methods.

Makes sense, and now that Warp is open sourced, community can carry some of the workload. Thanks.

**Comment by L0GYKAL** [comment]
@oz-agent please draft the spec.
